web/modules/custom/farm_nfa/src/PlanPermissionsRouteChecker.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace Drupal\farm_nfa;
+
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Routing\Access\AccessInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\user\Entity\Role;
+
+/**
+ * Checks access for displaying plan permissions route.
+ */
+class PlanPermissionsRouteChecker implements AccessInterface {
+
+  /**
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   Run access checks for this account.
+   */
+  public function access(AccountInterface $account, $plan) {
+    // Load the entity by ID.
+    $entity = \Drupal::entityTypeManager()->getStorage('plan')->load($plan);
+    $users = $entity->get('user')->referencedEntities();
+    $user_ids = [];
+    $roles = $account->getRoles();
+    $all_permissions = [];
+
+    foreach ($roles as $role_id) {
+      $role = Role::load($role_id);
+      if ($role) {
+        $permissions = $role->getPermissions();
+        $all_permissions = array_merge($all_permissions, $permissions);
+      }
+    }
+    foreach ($users as $user) {
+      $user_ids[] = $user->id();
+    }
+    foreach ($all_permissions as $permission) {
+      $can_access_plan_permissions = $permission == 'administer plans' ||
+        ($permission == 'grant edit on plans to other users' && in_array($account->id(), $user_ids));
+      if ($can_access_plan_permissions) {
+        return AccessResult::allowed();
+      }
+    }
+    return AccessResult::forbidden();
```

Comment:
Wondering if this should be neutral instead?

Reply:
@pcambra can you please give some more context around, what you meant by neutral here ?